libbeat/common/kubernetes/watcher.go

```diff
@@ -60,237 +61,228 @@ type WatchOptions struct {
 	case *StatefulSet:
-		list := &appsv1.StatefulSetList{}
-		w.resourceList = list
-		w.k8sResourceFactory = func() k8s.Resource { return &appsv1.StatefulSet{} }
-		w.items = func() []k8s.Resource {
-			rs := make([]k8s.Resource, 0, len(list.Items))
-			for _, item := range list.Items {
-				rs = append(rs, item)
-			}
-			return rs
+		ss := client.AppsV1().ReplicaSets(opts.Namespace)
+		listwatch = &cache.ListWatch{
+			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+				return ss.List(options)
+			},
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				return ss.Watch(options)
+			},
 		}
+
+		objType = "statefulset"
 	default:
 		return nil, fmt.Errorf("unsupported resource type for watching %T", resource)
 	}
-	return w, nil
-}
-
-func (w *watcher) AddEventHandler(h ResourceEventHandler) {
-	w.handler = h
-}
-
-func (w *watcher) buildOpts() []k8s.Option {
-	options := []k8s.Option{k8s.ResourceVersion(w.lastResourceVersion)}
-	if w.options.Node != "" {
-		options = append(options, filterByNode(w.options.Node))
-	}
-	return options
-}
-
-func (w *watcher) sync() error {
-	ctx, cancel := context.WithTimeout(w.ctx, w.options.SyncTimeout)
-	defer cancel()
 
-	logp.Info("kubernetes: Performing a resource sync for %T", w.resourceList)
-	err := w.client.List(ctx, w.options.Namespace, w.resourceList, w.buildOpts()...)
-	if err != nil {
-		logp.Err("kubernetes: Performing a resource sync err %s for %T", err.Error(), w.resourceList)
-		return err
-	}
+	informer = cache.NewSharedInformer(listwatch, resource, opts.SyncTimeout)
+	store = informer.GetStore()
+	queue = workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), objType)
+	ctx, cancel := context.WithCancel(context.Background())
 
-	w.logger.Debugf("Got %v items from the resource sync", len(w.items()))
-	for _, item := range w.items() {
-		w.onAdd(item)
+	w := &watcher{
+		client:   client,
+		informer: informer,
+		store:    store,
+		queue:    queue,
+		ctx:      ctx,
+		stop:     cancel,
+		logger:   logp.NewLogger("kubernetes"),
 	}
 
-	w.logger.Debugf("Done syncing %v items from the resource sync", len(w.items()))
-	// Store last version
-	w.lastResourceVersion = w.resourceList.GetMetadata().GetResourceVersion()
+	w.informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(o interface{}) {
+			w.enqueue(o, add)
+		},
+		DeleteFunc: func(o interface{}) {
+			w.enqueue(o, delete)
+		},
+		// TODO: When we move to one watcher per object model, we should check resource verion of
+		// TODO: the old vs new and avoid reconciliation during relisting.
+		UpdateFunc: func(_, o interface{}) {
+			w.enqueue(o, update)
+		},
+	})
 
-	logp.Info("kubernetes: %s", "Resource sync done")
-	return nil
+	return w, nil
 }
 
-func (w *watcher) onAdd(obj Resource) {
-	w.handler.OnAdd(obj)
-}
+func (w *watcher) enqueue(obj interface{}, state string) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
```

Comment:
what is this line for?

Reply:
this just makes sure that it gets the namespace/name combination only for resources where state is not `Unknown`